File: utils/src/main/java/ai/elimu/content_provider/utils/converter/CursorToLetterSoundGsonConverter.java
package ai.elimu.content_provider.utils.converter;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.util.Log;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import ai.elimu.model.v2.gson.content.LetterGson;
import ai.elimu.model.v2.gson.content.LetterSoundGson;
import ai.elimu.model.v2.gson.content.SoundGson;

public class CursorToLetterSoundGsonConverter {

    public static LetterSoundGson getLetterSoundGson(Cursor cursor, Context context, String contentProviderApplicationId) {
        Log.i(CursorToLetterSoundGsonConverter.class.getName(), "getLetterSoundGson");

        Log.i(CursorToLetterSoundGsonConverter.class.getName(), "Arrays.toString(cursor.getColumnNames()): " + Arrays.toString(cursor.getColumnNames()));

        int columnIndexId = cursor.getColumnIndex("id");
        Long id = cursor.getLong(columnIndexId);
        Log.i(CursorToLetterSoundGsonConverter.class.getName(), "id: " + id);

        int columnIndexRevisionNumber = cursor.getColumnIndex("revisionNumber");
        Integer revisionNumber = cursor.getInt(columnIndexRevisionNumber);
        Log.i(CursorToLetterSoundGsonConverter.class.getName(), "revisionNumber: " + revisionNumber);

        int columnIndexUsageCount = cursor.getColumnIndex("usageCount");
        Integer usageCount = cursor.getInt(columnIndexUsageCount);
        Log.i(CursorToLetterSoundGsonConverter.class.getName(), "usageCount: " + usageCount);

        List<LetterGson> letterGsons = null;
        Uri lettersUri = Uri.parse("content://" + contentProviderApplicationId + ".provider.letter_provider/letters/by-letter-sound-id/" + id);
        Log.i(CursorToLetterSoundGsonConverter.class.getName(), "lettersUri: " + lettersUri);
        Cursor lettersCursor = context.getContentResolver().query(lettersUri, null, null, null, null);
        if (lettersCursor == null) {
            Log.e(CursorToLetterSoundGsonConverter.class.getName(), "lettersCursor == null");
            Toast.makeText(context, "lettersCursor == null", Toast.LENGTH_LONG).show();
        } else {
            Log.i(CursorToLetterSoundGsonConverter.class.getName(), "lettersCursor.getCount(): " + lettersCursor.getCount());

            letterGsons = new ArrayList<>();

            boolean isLast = false;
            while (!isLast) {
                lettersCursor.moveToNext();

                // Convert from Room to Gson
                LetterGson letterGson = CursorToLetterGsonConverter.getLetterGson(lettersCursor);
                letterGsons.add(letterGson);

                isLast = lettersCursor.isLast();
            }

            lettersCursor.close();
            Log.i(CursorToLetterSoundGsonConverter.class.getName(), "lettersCursor.isClosed(): " + lettersCursor.isClosed());
        }

        List<SoundGson> soundGsons = null;
        Uri soundsUri = Uri.parse("content://" + contentProviderApplicationId + ".provider.sound_provider/sounds/by-letter-sound-id/" + id);
        Log.i(CursorToLetterSoundGsonConverter.class.getName(), "soundsUri: " + soundsUri);
        Cursor soundsCursor = context.getContentResolver().query(soundsUri, null, null, null, null);
        if (soundsCursor == null) {
            Log.e(CursorToLetterSoundGsonConverter.class.getName(), "soundsCursor == null");
            Toast.makeText(context, "soundsCursor == null", Toast.LENGTH_LONG).show();
        } else {
            Log.i(CursorToLetterSoundGsonConverter.class.getName(), "soundsCursor.getCount(): " + soundsCursor.getCount());

            soundGsons = new ArrayList<>();

            boolean isLast = false;
            while (!isLast) {
                soundsCursor.moveToNext();

                // Convert from Room to Gson
                SoundGson soundGson = CursorToSoundGsonConverter.getSoundGson(soundsCursor);
                soundGsons.add(soundGson);

                isLast = soundsCursor.isLast();
            }

            soundsCursor.close();
            Log.i(CursorToLetterSoundGsonConverter.class.getName(), "soundsCursor.isClosed(): " + soundsCursor.isClosed());
        }

        LetterSoundGson letterSound = new LetterSoundGson();
        letterSound.setId(id);
        letterSound.setRevisionNumber(revisionNumber);
        letterSound.setUsageCount(usageCount);
        letterSound.setLetters(letterGsons);
        letterSound.setSounds(soundGsons);

        return letterSound;
    }
}

File: utils/src/main/java/ai/elimu/content_provider/utils/converter/CursorToLetterSoundGsonConverter.kt
package ai.elimu.content_provider.utils.converter

import ai.elimu.content_provider.utils.converter.CursorToLetterGsonConverter.getLetterGson
import ai.elimu.content_provider.utils.converter.CursorToSoundGsonConverter.getSoundGson
import ai.elimu.model.v2.gson.content.LetterGson
import ai.elimu.model.v2.gson.content.LetterSoundGson
import ai.elimu.model.v2.gson.content.SoundGson
import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.util.Log
import android.widget.Toast
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

object CursorToLetterSoundGsonConverter {

    private const val TAG = "CursorToLetterSoundGsonConverter"
    private val mainScope by lazy { CoroutineScope(Dispatchers.Main) }

    fun getLetterSoundGson(
        cursor: Cursor,
        context: Context,
        contentProviderApplicationId: String
    ): LetterSoundGson {
        Log.i(TAG, "getLetterSoundGson")

        Log.i(TAG,
            "Arrays.toString(cursor.getColumnNames()): " + cursor.columnNames.contentToString())

        val columnIndexId = cursor.getColumnIndex("id")
        val id = cursor.getLong(columnIndexId)
        Log.i(TAG, "id: $id")

        val columnIndexRevisionNumber = cursor.getColumnIndex("revisionNumber")
        val revisionNumber = cursor.getInt(columnIndexRevisionNumber)
        Log.i(TAG, "revisionNumber: $revisionNumber")

        val columnIndexUsageCount = cursor.getColumnIndex("usageCount")
        val usageCount = cursor.getInt(columnIndexUsageCount)
        Log.i(TAG, "usageCount: $usageCount")

        val letterGsons = mutableListOf<LetterGson>()
        val lettersUri =
            Uri.parse("content://$contentProviderApplicationId.provider.letter_provider/letters/by-letter-sound-id/$id")
        Log.i(TAG, "lettersUri: $lettersUri")
        val lettersCursor = context.contentResolver.query(lettersUri, null, null, null, null)
        if (lettersCursor == null) {
            Log.e(TAG, "lettersCursor == null")
            mainScope.launch {
                Toast.makeText(context, "lettersCursor == null", Toast.LENGTH_LONG).show()
            }
        } else {
            Log.i(TAG, "lettersCursor.getCount(): " + lettersCursor.count)

            var isLast = false
            while (!isLast) {
                lettersCursor.moveToNext()

                // Convert from Room to Gson
                val letterGson = getLetterGson(lettersCursor)
                letterGsons.add(letterGson)

                isLast = lettersCursor.isLast
            }

            lettersCursor.close()
            Log.i(TAG, "lettersCursor.isClosed(): " + lettersCursor.isClosed)
        }

        val soundGsons = mutableListOf<SoundGson>()
        val soundsUri =
            Uri.parse("content://$contentProviderApplicationId.provider.sound_provider/sounds/by-letter-sound-id/$id")
        Log.i(TAG, "soundsUri: $soundsUri")
        val soundsCursor = context.contentResolver.query(soundsUri, null, null, null, null)
        if (soundsCursor == null) {
            Log.e(TAG, "soundsCursor == null")
            mainScope.launch {
                Toast.makeText(context, "soundsCursor == null", Toast.LENGTH_LONG).show()
            }
        } else {
            Log.i(TAG, "soundsCursor.getCount(): " + soundsCursor.count)

            var isLast = false
            while (!isLast) {
                soundsCursor.moveToNext()

                // Convert from Room to Gson
                val soundGson = getSoundGson(soundsCursor)
                soundGsons.add(soundGson)

                isLast = soundsCursor.isLast
            }

            soundsCursor.close()
            Log.i(TAG, "soundsCursor.isClosed(): " + soundsCursor.isClosed)
        }

        return LetterSoundGson().apply {
            this.id = id
            this.revisionNumber = revisionNumber
            this.usageCount = usageCount
            this.letters = letterGsons
            this.sounds = soundGsons
        }
    }
}
